File: s3/Ruby/lib/s3client.rb
# frozen_string_literal: true

# Copyright Core DF
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'aws-sdk-s3'
require_relative '../../../http/Ruby/lib/result'

module S3client
  module_function

  def aws_client
    region = ENV['AWS_REGION'] || ENV['AWS_DEFAULT_REGION'] || 'us-east-1'
    opts = { region: region }
    endpoint = ENV['S3_ENDPOINT_URL']
    if endpoint && !endpoint.empty?
      opts[:endpoint] = endpoint
      opts[:force_path_style] = true
    end
    Aws::S3::Client.new(opts)
  end

  def bucket(explicit = nil)
    explicit && !explicit.empty? ? explicit : ENV['S3_BUCKET']
  end

  def Init
    if ENV['AWS_ACCESS_KEY_ID'].to_s.empty? && ENV['AWS_PROFILE'].to_s.empty?
      return CoreautoResult.missing_env('AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE')
    end
    return CoreautoResult.missing_env('S3_BUCKET (or pass bucket per call)') if ENV['S3_BUCKET'].to_s.empty?

    { status_code: 200 }
  end

  def GetObject(key, bucket_name = nil)
    b = bucket(bucket_name)
    return CoreautoResult.missing_env('S3_BUCKET') if b.to_s.empty?

    resp = aws_client.get_object(bucket: b, key: key)
    raw = resp.body.read
    text = raw.dup.force_encoding(Encoding::UTF_8)
    content = text.valid_encoding? ? text : raw
    { status_code: 200, content: content }
  rescue StandardError => e
    CoreautoResult.transport_error(e.message)
  end

  def PutObject(key, content, bucket_name = nil)
    b = bucket(bucket_name)
    return CoreautoResult.missing_env('S3_BUCKET') if b.to_s.empty?

    aws_client.put_object(bucket: b, key: key, body: content)
    { status_code: 200 }
  rescue StandardError => e
    CoreautoResult.transport_error(e.message)
  end

  def ListObjects(prefix = '', bucket_name = nil)
    b = bucket(bucket_name)
    return CoreautoResult.missing_env('S3_BUCKET') if b.to_s.empty?

    resp = aws_client.list_objects_v2(bucket: b, prefix: prefix)
    keys = (resp.contents || []).map(&:key)
    { status_code: 200, keys: keys }
  rescue StandardError => e
    CoreautoResult.transport_error(e.message)
  end
end

File: s3/Rust/src/s3client.rs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::result::{missing_env, transport_error};
use aws_sdk_s3::primitives::ByteStream;
use aws_sdk_s3::Client;
use serde_json::{json, Value};
use std::env;
use std::str;

fn region() -> String {
    env::var("AWS_REGION")
        .or_else(|_| env::var("AWS_DEFAULT_REGION"))
        .unwrap_or_else(|_| "us-east-1".into())
}

fn bucket(explicit: Option<&str>) -> String {
    explicit
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .or_else(|| env::var("S3_BUCKET").ok())
        .unwrap_or_default()
}

async fn aws_client() -> Result<Client, String> {
    let shared = aws_config::defaults(aws_config::BehaviorVersion::latest())
        .region(aws_config::Region::new(region()))
        .load()
        .await;
    let mut s3_builder = aws_sdk_s3::config::Builder::from(&shared);
    if let Ok(endpoint) = env::var("S3_ENDPOINT_URL") {
        if !endpoint.is_empty() {
            s3_builder = s3_builder
                .endpoint_url(endpoint)
                .force_path_style(true);
        }
    }
    Ok(Client::from_conf(s3_builder.build()))
}

fn block_on<F: std::future::Future<Output = Value>>(f: F) -> Value {
    match tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
    {
        Ok(rt) => rt.block_on(f),
        Err(e) => transport_error(&e.to_string()),
    }
}

pub fn init() -> Value {
    if env::var("AWS_ACCESS_KEY_ID").unwrap_or_default().is_empty()
        && env::var("AWS_PROFILE").unwrap_or_default().is_empty()
    {
        return missing_env("AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE");
    }
    if env::var("S3_BUCKET").unwrap_or_default().is_empty() {
        return missing_env("S3_BUCKET (or pass bucket per call)");
    }
    json!({ "status_code": 200 })
}

pub fn get_object(key: &str, bucket_name: Option<&str>) -> Value {
    let b = bucket(bucket_name);
    if b.is_empty() {
        return missing_env("S3_BUCKET");
    }
    block_on(async move {
        let client = match aws_client().await {
            Ok(c) => c,
            Err(e) => return transport_error(&e),
        };
        let resp = match client.get_object().bucket(&b).key(key).send().await {
            Ok(r) => r,
            Err(e) => return transport_error(&e.to_string()),
        };
        let data = match resp.body.collect().await {
            Ok(b) => b.into_bytes().to_vec(),
            Err(e) => return transport_error(&e.to_string()),
        };
        let content: Value = match str::from_utf8(&data) {
            Ok(s) => Value::String(s.to_string()),
            Err(_) => json!(data),
        };
        json!({ "status_code": 200, "content": content })
    })
}

pub fn put_object(key: &str, content: &str, bucket_name: Option<&str>) -> Value {
    let b = bucket(bucket_name);
    if b.is_empty() {
        return missing_env("S3_BUCKET");
    }
    block_on(async move {
        let client = match aws_client().await {
            Ok(c) => c,
            Err(e) => return transport_error(&e),
        };
        match client
            .put_object()
            .bucket(&b)
            .key(key)
            .body(ByteStream::from(content.as_bytes().to_vec()))
            .send()
            .await
        {
            Ok(_) => json!({ "status_code": 200 }),
            Err(e) => transport_error(&e.to_string()),
        }
    })
}

pub fn list_objects(prefix: &str, bucket_name: Option<&str>) -> Value {
    let b = bucket(bucket_name);
    if b.is_empty() {
        return missing_env("S3_BUCKET");
    }
    let prefix = prefix.to_string();
    block_on(async move {
        let client = match aws_client().await {
            Ok(c) => c,
            Err(e) => return transport_error(&e),
        };
        let resp = match client
            .list_objects_v2()
            .bucket(&b)
            .prefix(&prefix)
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => return transport_error(&e.to_string()),
        };
        let keys: Vec<String> = resp
            .contents()
            .iter()
            .filter_map(|o| o.key().map(str::to_string))
            .collect();
        json!({ "status_code": 200, "keys": keys })
    })
}
